octokit/client.py

```diff
@@ -28,8 +29,43 @@ class Client(Resource):
     'mastahyeti'
   """
 
-  def __init__(self, api_endpoint='https://api.github.com', **kwargs):
-    self.session = Session(**kwargs)
+  def __init__(self, session=requests.Session(), api_endpoint='https://api.github.com', **kwargs):
+    self.session = session
     self.url = api_endpoint
     self.schema = {}
     self.name = 'Client'
+    self._rate_limit = _RateLimit()
+    self.auto_paginate = True
+
+    self.session.hooks = dict(response=self.response_callback)
+    for key in kwargs:
+      setattr(self.session, key, kwargs[key])
+
+  def response_callback(self, r, *args, **kwargs):
+    self.last_response = r
+    # TODO (howei): perhaps we could auto-paginate requests here
```

Comment:
Yep. We could also address one of your TODOs in agent by handling HTTP response status codes here instead of in Resource (since it's octokit specific).